### PR TITLE
feat(@clayui/data-provider): pass formatted variables when link is a function

### PR DESCRIPTION
**Pull request by matuzalemsteles**

### CODE REVIEW

File: packages/clay-data-provider/src/__tests__/index.tsx [hunk 1]
```diff
@@ -267,6 +281,25 @@ describe('ClayDataProvider', () => {
 		);
 	});
 
+	it('call clay.data with search params and link being a function', async () => {
+		fetchMock.mockResponse(JSON.stringify({title: 'Foo'}));
+
+		render(
+			<DataProvider
+				link={variables => fetch(`https://clay.data/?${variables}`)}
```
Comment (bryceosterhaus) [inline packages/clay-data-provider/src/__tests__/index.tsx]: How come we pass in the variables via a prop, only to then consume it via link? Seems like the user should just form the link url themselves. I saw the comment below in `warning()` but still wasn't quite clear on it.

```js
link={variables => fetch(`https://clay.data/?name=bar`)}
```

Comment (matuzalemsteles) [reply packages/clay-data-provider/src/__tests__/index.tsx]: This is just to add support for using variables for when the `link` prop is a function, if you set the URL only in your fetch, you lose the benefits of `variables` functionality, for example: new fetch when `variables` change, improve cache match... This makes some improvements to #2261 use cases.

You can even set the URL in fetch method and use variables just to have that benefit and in your fetch URL, but you'd better declare it in one place and encourage people to use the API.

Comment (bryceosterhaus) [reply packages/clay-data-provider/src/__tests__/index.tsx]: ah i see. Just wanted to make sure I was understanding it correctly.

File: packages/clay-data-provider/src/types.ts [hunk 2]
```diff
@@ -36,14 +36,16 @@ export enum FetchPolicy {
...
 
-export type TLink = Function | string;
-
 export interface IVariables {
 	[propName: string]: any;
 }
 
 export type TVariables = IVariables | null;
 
+export type TLinkFunction = (variables: string) => Promise<any>;
```
Comment (wincent) [inline packages/clay-data-provider/src/types.ts]: Thanks for improving the type here!

**Pull request merged**